lib/SILOptimizer/Utils/Existential.cpp

```diff
@@ -111,6 +111,19 @@ static SILInstruction *getStackInitInst(SILValue allocStackAddr,
       }
       continue;
     }
+    if (auto *store = dyn_cast<StoreInst>(User)) {
+      if (store->getDest() == allocStackAddr) {
+        if (SingleWrite)
+          return nullptr;
+        SingleWrite = store;
+        // When we support OSSA here, we need to insert a new copy of the value
+        // before `store` (and make sure that the copy is destroyed when
+        // replacing the apply operand).
```

Comment:
My previous comment might have been misleading because I didn't realize we're inserting a new destroy. In OSSA, the store will do an implicit copy. But, currently, you need to copy the stored value, probably using a retain_value before the store.

To keep the ownership clean what probably ends up happening is this *I think*:

We have an original existential argument value and a new concrete argument value.

The original existential argument value is removed from a "consuming" call operand, and a new destroy_addr of that value is generated in its place (right after the call).

The copy of the new concrete argument value is emitted. That copy is then consumed by the call argument that it is assigned to.

That will be fairly straightforward to update for OSSA later.

Reply:
Here's a question I've been wanting to ask for a while: what's the relationship between OSSA and reference counting? Is it handled differently in OSSA?

Thanks for laying it all out. That mostly makes sense except for the part about adding a new `destroy_addr` instruction. Currently, the destroy_addr is for the _new_ value (the "tempArg") are you saying it should instead (or in addition) be for the old argument? 

Again, thank you for all the help and information you've given me in this patch. I'm very grateful.

Reply:
OSSA is a more strict form of SIL. That's all. It means that each SSA value has its own ownership. So you can't arbitrarily retain/release values and hope they all balance out. A value can only be consumed once (by a destroy_value or owned argument) and cannot be used after it is consumed.

It looks to me like the new destroy at the call site is only created when a `ConcreteArgumentCopy` was generated for the argument. Since ConcreateArgumentCopy always copies/retains, the new destroy will balance that out.

There's another issue I was worried about, which is a totally separate issue from ConcreteArgumentCopy: getStackInitInst will discover the source of the store instruction. We use that as the new ConcreteValue. But, we don't set `isConcreteValueCopied`. So, for stored values, we never check `canReplaceCopiedArg`. The purpose of canReplaceCopiedArg is to ensure that the concrete value is "available" at the call site. So you could have:

```
address = alloc_stack
retain concreteValue // copy to stack
store concreteValue to address
release concreteValue // concreteValue out of scope
apply (@guaranteed address)
use address
destroy address
```

Optimizes to:
```
address = alloc_stack
retain concreteValue // copy to stack
store concreteValue to address
release concreteValue // no longer makes sense
apply (@guaranteed concreteValue)
use address
destroy address
```

That's kind of an abuse of SIL, but it's fine for non-OSSA, and in
fact optimal because ARCCodeMotion will turn it into this:

```
address = alloc_stack
store concreteValue to address
apply (@guaranteed concreteValue)
use address
destroy address
```


But with OSSA it will be:

```
address = alloc_stack
store [init] concreteValue to address
destroy concreteValue
apply (address)
use address
destroy address
```

Optimizes to:
```
address = alloc_stack
copied = copy_value concreteValue
store [init] concreteValue to address
destroy concreteValue
apply (concreteValue)
destroy copied
use address
destroy address
```

I think this PR is fine as-is since you've added a comment for how to handle OSSA later.

Reply:
Interesting. Really good to know, thank you for taking the time to explain. I'm happy to work on updating more parts of the codebase to use OSSA if you or someone else can help me get started (the knowledge transfer may be more effort than it's worth, though).